### PR TITLE
Implement peer blacklist

### PR DESCRIPTION
**Pull request by vyzo**
Adds a peer blacklist, which is checked before doing any processing on an incoming message.

### CODE REVIEW

**Comment by vyzo** [comment]
sorry for the force pushing mess, i was trying to get the test coverage right... it should be ready for review now.

**Comment by vyzo** [comment]
there is a hole currently in that the incoming stream is not tracked; this allows a blacklisted peer to replay old valid messages from other peers.

**Comment by vyzo** [comment]
actually it's ok, we do the blacklist check on the node that forwarded the message. but we should also check the message origin as well, as it might be blacklisted.

**Comment by vyzo** [comment]
the issue is that the validator can't see the peer that forwarded the message, so it can't make blacklist decisions for that peer.

**Comment by vyzo** [comment]
follow up in #150.

File: pubsub.go [hunk 1]
```diff
@@ -179,6 +183,8 @@ func NewPubSub(ctx context.Context, h host.Host, rt PubSubRouter, opts ...Option
 		topics:           make(map[string]map[peer.ID]struct{}),
 		peers:            make(map[peer.ID]chan *RPC),
 		topicVals:        make(map[string]*topicVal),
+		blacklist:        make(map[peer.ID]struct{}),
```
Comment (Stebalien) [inline pubsub.go]: May want to use an LRU. This could turn into a DoS vector.

**Review by Stebalien — COMMENTED** [review]
We should probably use an LRU to prevent bad peers from forcing us to leak memory.

File: pubsub.go (hunk 1 above)
Comment (vyzo) [reply pubsub.go]: Good point. As we discussed with why, let's make it an interface the user can pass with a default implementation using a map.

**Comment by vyzo** [comment]
Added a blacklist type, with map and lru cache backed implementations.
The blacklist can be configured with the `WithBlacklist` option.

**Review by Stebalien — APPROVED** [review]

**Pull request merged**